extensions/standard-processors/processors/GenerateFlowFile.cpp

```diff
@@ -57,58 +56,78 @@ void generateData(std::vector<char>& data, bool textData = false) {
+  } else {
+    if (is_text)
+      return Mode::NotUniqueText;
+    else
+      return Mode::NotUniqueByte;
   }
-  if (context.getProperty(UniqueFlowFiles.name, uniqueFlowFile_)) {
-    logger_->log_trace("Unique Flow files is configured to be {}", uniqueFlowFile_);
+}
+
+void GenerateFlowFile::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
+  bool is_text = context.getProperty<std::string>(DataFormat)
+      | utils::transform([](const std::string& data_format) { return data_format == DATA_FORMAT_TEXT;})
+      | utils::valueOrElse([]() {return false;});
+  bool is_unique = context.getProperty<bool>(UniqueFlowFiles) | utils::valueOrElse([] { return true; });
+
+  auto custom_text_without_evaluation = context.getProperty(CustomText);
+  bool has_custom_text = custom_text_without_evaluation.has_value() && !custom_text_without_evaluation->empty();
+
+  context.getProperty(FileSize, file_size_);
+  context.getProperty(BatchSize, batch_size_);
+
+  mode_ = getMode(is_unique, is_text, has_custom_text, file_size_);
+
+  if (!isUnique(mode_)) {
+    non_unique_data_.resize(gsl::narrow<size_t>(file_size_));
+    generateData(non_unique_data_, isText(mode_));
   }
 
+  logger_->log_trace("GenerateFlowFile is configured in {} mode", magic_enum::enum_name(mode_));
+  if (mode_ != Mode::CustomText && has_custom_text)
+    logger_->log_warn("Custom Text property is set, but not used!");
+}
+
+// The custom text has to be reevaluated once per batch
+void GenerateFlowFile::refreshNonUniqueData(core::ProcessContext& context) {
+  if (mode_ != Mode::CustomText)
+    return;
   std::string custom_text;
   context.getProperty(CustomText, custom_text, nullptr);
-  if (!custom_text.empty()) {
-    if (textData_ && !uniqueFlowFile_) {
-      data_.assign(custom_text.begin(), custom_text.end());
-      return;
-    } else {
-      logger_->log_warn("Custom Text property is set, but not used!");
-    }
-  }
-
-  if (!uniqueFlowFile_) {
-    data_.resize(gsl::narrow<size_t>(fileSize_));
-    generateData(data_, textData_);
-  }
+  non_unique_data_.assign(custom_text.begin(), custom_text.end());
 }
 
-void GenerateFlowFile::onTrigger(core::ProcessContext&, core::ProcessSession& session) {
-  for (uint64_t i = 0; i < batchSize_; i++) {
+void GenerateFlowFile::onTrigger(core::ProcessContext& context, core::ProcessSession& session) {
+  refreshNonUniqueData(context);
+  for (uint64_t i = 0; i < batch_size_; i++) {
     // For each batch
-    std::shared_ptr<core::FlowFile> flowFile = session.create();
-    if (!flowFile) {
+    std::shared_ptr<core::FlowFile> flow_file = session.create();
+    if (!flow_file) {
       logger_->log_error("Failed to create flowfile!");
       return;
     }
-    if (uniqueFlowFile_) {
-      std::vector<char> data(gsl::narrow<size_t>(fileSize_));
-      if (fileSize_ > 0) {
-        generateData(data, textData_);
-      }
-      session.writeBuffer(flowFile, data);
+    if (mode_ == Mode::Empty) {
+      // noop
```

Comment:
Shouldn't we emit an empty flow file here?

Reply:
we do, the creation is in line 116 and the transfer is after all these ifs in line 130, We only skip the modification of the created empty flowfile